sentry_streams/sentry_streams/adapters/arroyo/broadcaster.py

```diff
@@ -0,0 +1,130 @@
+    ) -> None:
+        self.__next_step = next_step
+        self.__route = route
+        self.__downstream_branches = downstream_branches
+        # If we get MessageRejected from the next step, we put the pending messages here
+        self.__failed_msgs: MutableMapping[Messageidentifier, Message[RoutedValue]] = {}
+
+    def __retry_failed_msg(
+        self, msg: Message[RoutedValue], msg_identifier: Messageidentifier
+    ) -> None:
+        """
+        Retries a message which got MessageRejected previously.
+        """
+        self.__next_step.submit(msg)
+        del self.__failed_msgs[msg_identifier]
+
+    def __submit_to_next_step(
+        self, msg: Message[RoutedValue], msg_identifier: Messageidentifier
+    ) -> None:
+        """
+        Submits to the next step.
+        If the next step raises MessageRejected, this records the failed message
+        and raises MessageRejected to propagate the error.
+        """
+        try:
+            self.__next_step.submit(msg)
+        except MessageRejected:
+            self.__failed_msgs[msg_identifier] = msg
+            raise MessageRejected()
+
+    def __handle_submit(self, msg: Message[RoutedValue]) -> None:
+        msg_identifier = Messageidentifier(
+            route=msg.value.payload.route,
+            committable=msg.value.committable,
+        )
+        if msg_identifier in self.__failed_msgs:
+            self.__retry_failed_msg(msg, msg_identifier)
+        else:
+            self.__submit_to_next_step(msg, msg_identifier)
+
+    def submit(self, message: Message[Union[FilteredPayload, RoutedValue]]) -> None:
+        # if any downstream branch raises MessageRejected, we need to propagate
+        # the error back to the previous step
+        raise_message_rejected = False
+        if (
+            isinstance(message.value.payload, RoutedValue)
+            and message.value.payload.route == self.__route
+        ):
+            for branch in self.__downstream_branches:
+                msg_copy = cast(Message[RoutedValue], deepcopy(message))
+                copy_payload = msg_copy.value.payload
+                routed_copy = Message(
+                    Value(
+                        committable=msg_copy.value.committable,
+                        timestamp=msg_copy.value.timestamp,
+                        payload=RoutedValue(
+                            route=Route(
+                                source=copy_payload.route.source,
+                                waypoints=[*copy_payload.route.waypoints, branch],
+                            ),
+                            payload=copy_payload.payload,
+                        ),
+                    )
+                )
+                try:
+                    self.__handle_submit(routed_copy)
+                except MessageRejected:
+                    # Raise a single MessageRejected at the end of the loop
+                    raise_message_rejected = True
```

Comment:
You do not need to do this.
Just raise the exception at the first failure.
So you can only have one failed message